drv/stm32xx-i2c/src/lib.rs

```diff
@@ -466,6 +492,62 @@ impl I2cController<'_> {
         Ok(())
     }
 
+    ///
+    /// A routine to panic.  This should not be called merely because something
+    /// has gone wrong with a device (which should rather be indicated by
+    /// returning an error and resetting the controller if/as needed), but with
+    /// the controller itself.
+    ///
+    fn panic(&self) -> ! {
+        let i2c = self.registers;
+        let tgr = &i2c.timingr;
+        let tor = &i2c.timeoutr;
+
+        ringbuf_entry!(Trace::Panic(Register::CR1, i2c.cr1.read().bits()));
+        ringbuf_entry!(Trace::Panic(Register::CR2, i2c.cr2.read().bits()));
+        ringbuf_entry!(Trace::Panic(Register::OAR1, i2c.oar1.read().bits()));
+        ringbuf_entry!(Trace::Panic(Register::OAR2, i2c.oar2.read().bits()));
+        ringbuf_entry!(Trace::Panic(Register::TIMINGR, tgr.read().bits()));
+        ringbuf_entry!(Trace::Panic(Register::TIMEOUTR, tor.read().bits()));
+        ringbuf_entry!(Trace::Panic(Register::ISR, i2c.isr.read().bits()));
+        ringbuf_entry!(Trace::Panic(Register::PECR, i2c.pecr.read().bits()));
```

Comment:
for my own education, what's the motivation for recording each register as a separate ringbuf entry rather than one entry with all of them? something like this:
```rust
enum Trace {
    // ...
    Panic {
         cr1: u32,
         cr2: u32,
         oar1: u32,
         oar2: u32,
         // ...
    },
    // ...
}
```
it seems like recording each register as its own entry occupies more slots in the ringbuf, potentially pushing stuff off the end that might be useful for diagnosing the panic?

i'm guessing that the motivation for this is that keeping the entry type itself small makes the ringbuf as a whole smaller, and that's worth the risk of potentially losing earlier data? is that correct?

Reply:
Yes, all about keeping the entry size small to maximize the number of entries. There are other cases (e.g., recording the MAX5970 state) where there isn't much variance in the size of the entry, so we record all of the state in a single entry.

Reply:
cool, that's what i figured, but i wanted to check. thanks!